### PR TITLE
Fix AA-366. Add zero reference date for heart rate data.

### PR DESCRIPTION
**Pull request by nategbrown9**

### CODE REVIEW

**Comment by liujoshua** [comment]
IMO it would be okay to implement `uptime` either as 1) `timestamp` in seconds or 2) the unknown-zero camera frame timestamp in seconds
@syoung-smallwisdom does that sound right?

**Comment by syoung-smallwisdom** [comment]
Ideally, uptime should be something that is the same value with the same meaning across different files. On iOS, this is "clock time" b/c that's easy to get and all other "timestamps" are derived from it.

I intentionally wrote the interface to *require* `uptime` but not to require the `timestampDate` or `timestamp` b/c ResearchKit was completely inconsistent in the use of the term timestamp. Mostly, the "timestamp" was the time since the recorder was started, but sometimes it was system uptime and sometimes it was an `Instance` timestamp *date*.

So I decided that both `timestampDate` and `timestamp` are optional (except for marking the start of the recording which includes `timestampDate`, `stepPath`, `uptime`, and `timestamp=0`).

I'm not sure how best to reconcile this. I would have to do more spelunking into the code that is handling recording the acceleration and gyro, but `uptime` here should cross-reference to `uptime` from the motion sensors and be the same value.

```
long now = frame.getMetadata().getTimestampMillis();
```

Is the frame timestamp relative to clock time or to when the video was started?

**Comment by liujoshua** [comment]
My understanding is the timestamps for an image frame have an zero is different for each camera on the device and we don't know how to associate these zeroes to calendar time. So our `timestampDate` is an estimate but we are now preserving the time intervals between each `timestamp` in a JSON file (this PR fixes the jitter we were introducing).

@itismeghasyam do you rely on `uptime` having a consistent zero across the JSON files for an ActiveTask? Or would `timestamp` and `timestampDate` give you enough info?

**Comment by syoung-smallwisdom** [comment]
I'd like to fix this "correctly" where by that I mean that we are using a consistent terminology for all record samples and across both platforms.

While @itismeghasyam may not use this information to cross-reference between motion sensors and the video frames, we already have third-party developers who *are* using the iOS version of this.

**Comment by liujoshua** [comment]
We use system uptime where we have it (e.g. sensors). I think the best we can do is apply the approach we use for estimating image timestamp's relationship to calendar time in order to estimate image timestamp's relationship to system uptime

**Comment by syoung-smallwisdom** [comment]
What I mean is do this:

https://github.com/Sage-Bionetworks/CardiorespiratoryFitness-Android/pull/83

File: app/src/main/java/org/sagebase/crf/step/active/BpmRecorder.java [hunk 1]
```diff
@@ -224,6 +229,18 @@ public HeartBeatJsonWriter(BpmUpdateListener
...
         public void onHeartRateSampleDetected(HeartBeatSample sample) {
             bpmCalculator.calculateBpm(sample);
+
+            if (timestampZeroReference <= 0) {
+                // set timestamp reference, which timestamps are measured relative to
+                timestampZeroReference = sample.t;
+
+                Date timestampReferenceDate = new Date(System.currentTimeMillis());
+                mJsonObject.addProperty(TIMESTAMP_DATE_KEY,
+                        new SimpleDateFormat(FormatHelper.DATE_FORMAT_ISO_8601, Locale.getDefault())
```
[the remark this answers is not in the document]
Comment (liujoshua) [reply app/src/main/java/org/sagebase/crf/step/active/BpmRecorder.java]: `new Locale("en", "US", "POSIX")`

File: app/src/main/java/org/sagebase/crf/step/active/BpmRecorder.java [hunk 2]
```diff
@@ -224,8 +232,27 @@ public HeartBeatJsonWriter(BpmUpdateListener
...
+            if (timestampZeroReference < 0) {
+                // set timestamp reference, which timestamps are measured relative to
+                timestampZeroReference = sample.t;
+                uptimeZeroReference = System.nanoTime() * 1e-9;
+
+                Date timestampReferenceDate = new Date(System.currentTimeMillis());
+                // TODO: syoung 10/04/2018 Figure out how to set timestamp dates to "en_US_POSIX"
+                mJsonObject.addProperty(TIMESTAMP_DATE_KEY,
+                        new SimpleDateFormat(FormatHelper.DATE_FORMAT_ISO_8601, new Locale("en", "us", "POSIX"))
+                                .format(timestampReferenceDate));
```
Comment (liujoshua) [inline app/src/main/java/org/sagebase/crf/step/active/BpmRecorder.java]: @nategbrown9  Could you pleasea remove Shannon's TODO above?

We also should make this change in SensorRecorder, LocationRecorder, ConsentSignatureStepLayout, and FormatHelper.

**Pull request merged**